# **Data Bias Assignment: Testing Model Fairness for Hate Speech Tweets for two marginalized groups of American Society**

**Chinmayee Channuri**




# **Aim + Methodology:**

In this study, I aimed to utilize the Perspective API and test it with regards to social media hate speech against two marginalized groups of society, Asian Americans and African Americans. Hence, my sensitive attributes were "Against Asian-Americans" and "Against African Americans".

I collected a sample size of 50 tweets (comprising of tweets with hate speech for Asian Americans and African Americans), and marked them with binary identifiers, "Toxic" or "Non-Toxic" according to how I felt about them, or what my reaction was to them when I read them. These values became my expected values.

Then, I utilized the Perspective API to run each tweet text through it to receive a predicted value for the toxicity of the tweet, which was a number in the range of 0-1.

I loaded the CSV file with my data in it using Pandas into my Jupyter Notebook to analyze and apply the accuracy study to it.

The main transformation I performed for my dataset was when I added a seperate column, "Predicted Label" into the data frame to deem a tweet as toxic or non-toxic based on the predicted values from Perspective API.

Then, I analysed the Model Fairness from the Predictive Equity standpoint to measure and observe the accuracy performance for each category, Toxic v. Non-Toxic, for each ethnicity group, Asian Americans and African Americans.


# **Hypothesis:**

I hypothesize that there would be differences between the accuracies for the toxicity levels for hate speech against Asian Americans and African Americans to account for the dataset that Perspective API was potentially trained on.

# **API Call Code**

In [ ]:
from googleapiclient import discovery
import json

API_KEY = 'AIzaSyD336HZlTbzcDu59k3Vi9Mq14eT0qbBJp0'

client = discovery.build(
  "commentanalyzer",
  "v1alpha1",
  developerKey=API_KEY,
  discoveryServiceUrl="https://commentanalyzer.googleapis.com/$discovery/rest?version=v1alpha1",
  static_discovery=False,
)

analyze_request = {
  'comment': { 'text': "" },
  'requestedAttributes': {'TOXICITY': {}}
}

response = client.comments().analyze(body=analyze_request).execute()
print(json.dumps(response, indent=2))




# **Downloaded the table as a CSV and loaded it here via Numpy and Pandas**

In [6]:
import numpy as np
import pandas as pd

test_df = pd.read_csv("data-bias-racism-tweets-dataset.csv")
test_df

# change expecyted score to expected label
# add another column into the csv
# loop over preditced scores and create a list over the predicted label make it toxic vs non toxic

def create_label(score):
  if score > 0.5:
    return "Toxic"
  else:
    return "Non Toxic"

test_df['Predicted Label'] = test_df['Predicted Score'].apply(create_label)
test_df

Sensitive Attribute  \
0     Against Asian Americans   
1     Against Asian Americans   
2     Against Asian Americans   
3     Against Asian Americans   
4     Against Asian Americans   
5     Against Asian Americans   
6     Against Asian Americans   
7     Against Asian Americans   
8     Against Asian Americans   
9     Against Asian Americans   
10    Against Asian Americans   
11    Against Asian Americans   
12    Against Asian Americans   
13    Against Asian Americans   
14    Against Asian Americans   
15    Against Asian Americans   
16    Against Asian Americans   
17    Against Asian Americans   
18    Against Asian Americans   
19    Against Asian Americans   
20    Against Asian Americans   
21    Against Asian Americans   
22    Against Asian Americans   
23    Against Asian Americans   
24    Against Asian Americans   
25    Against Asian Americans   
26    Against Asian Americans   
27    Against Asian Americans   
28  Against African Americans   
29  Against African Americans   
30  Against African Americans   
31  Against African Americans   
32  Against African Americans   
33  Against African Americans   
34  Against African Americans   
35  Against African Americans   
36  Against African Americans   
37  Against African Americans   
38  Against African Americans   
39  Against African Americans   
40                        NaN   
41  Against African Americans   
42  Against African Americans   
43  Against African Americans   
44  Against African Americans   
45  Against African Americans   
46  Against African Americans   
47  Against African Americans   
48  Against African Americans   
49  Against African Americans   

                                                 Text Expected Score  \
0   #COVID19 so this all started cuz some chinks w...          Toxic   
1   Hey China thanks for literally fucking up the ...          Toxic   
2   Yeah no shit, it only took the mainstream medi...          Toxic   
3   @LeoVaradkar Why is he licking China’s boots. ...          Toxic   
4   If you’re just now learning that China fudges ...          Toxic   
5   CCP China doing all types of shady &amp; evil ...          Toxic   
6   @PockyParade Are you retarded? China lied and ...          Toxic   
7   President Trump was called a racist calling co...          Toxic   
8   #ChinaVirus when will these dumbasses learn fr...         Toxic    
9   I AGREE WITH @SenTomCotton THE CORONAVIRUS MOR...          Toxic   
10  @skaterjeongyeon @ti_xu_dao Obviously the majo...          Toxic   
11  China is going to kill America one way or anot...          Toxic   
12  You’re just mad because you maybe can’t use th...          Toxic   
13  coronavirus is not xenophobic or racist and wi...      Non Toxic   
14  STOP CALLING COVID-19 THE “CHINESE VIRUS” WHAT...      Non Toxic   
15  @SenKamalaHarris @GOPLeader Hey Kevin - Callin...      Non Toxic   
16  Being racist against the Chinese won't stop th...      Non Toxic   
17  Put a #STOP 🛑 to all this racism Asians have b...      Non Toxic   
18  STOP FUCKING CALLING IT THE 'CHINESE VIRUS' yo...      Non Toxic   
19  Why was it so normal for people to call me a c...      Non Toxic   
20  I am SICK 👏 OF 👏 IT 👏 AND SO IS EVERY ASIAN ou...      Non Toxic   
21  I am tweeting to Support #IAmNotAVirus hashtag...      Non Toxic   
22  STOP CALLING COVID-19 THE “CHINESE VIRUS” THE ...      Non Toxic   
23  Shame on you if you attack Asians over this vi...      Non Toxic   
24  My husband continues to deliberately call it t...      Non Toxic   
25  @gabewildau It's called preventative actions. ...      Non Toxic   
26  Discrimination is not okay no matter what the ...      Non Toxic   
27  We are not diseased. We are not scapegoats. We...      Non Toxic   
28  #BlackLivesMatter are a bunch of #terrorists, ...          Toxic   
29  BlackLivesMatter Wake Up you are Pawns Being u...          Toxic   
30  Three feral black teens kill elderly white man...         Toxic    
31  Black lives matter

#**Assign a row a binary factor deeming it either toxic, 1, or non-toxic, 0, for both expected score (Toxic/Non-Toxic) and predicted score (>0.5/<0.5) columns**

In [2]:
y_actual = [1 if y.strip() == 'Toxic' else 0 for y in test_df['Expected Score']]
y_predicted = [1 if y.strip() == 'Toxic' else 0 for y in test_df['Predicted Label']]

# **Measured Accuracy of Model via Pickle and SkLearn Machine Learning Models**

In [3]:
import pickle
from sklearn.metrics import accuracy_score

accuracy = accuracy_score(y_predicted, y_actual)


print (f"Accuracy of the classifier = {accuracy}")

Accuracy of the classifier = 0.66


#**Analyzing the Accuracy of the Classifier:**

The overall accuracy of the model is moderatly high, at 66%. But, further analysis must be conducted to understand if the model is able to take into consideration the semantics, mood,message, and overall tone of the tweet rather than just taking the words used at face value and mapping them to being toxic or non-toxic.

In [4]:
race_column = test_df["Sensitive Attribute"]
asian_indices = []
african_american_indices = []

for ele in range(len(race_column)):
    if race_column[ele] == "Against Asian Americans":
      asian_indices.append(ele)
    else:
      african_american_indices.append(ele)

y_actual_asian = [y_actual[i] for i in asian_indices]
y_predicted_asian = [y_predicted[i] for i in asian_indices]

y_actual_african_american = [y_actual[i] for i in african_american_indices]
y_predicted_african_american = [y_predicted[i] for i in african_american_indices]

print("Number of Asian indices:", len(asian_indices))
print("Number of African American indices:", len(african_american_indices))

print(y_actual_asian)
print(y_predicted_asian)
print(y_actual_african_american)
print(y_predicted_african_american)



Number of Asian indices: 28
Number of African American indices: 22
[1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]
[0, 1, 1, 1, 1, 1, 1, 0, 0, 0, 1, 0, 1, 0, 0, 1, 0, 0, 1, 0, 1, 0, 1, 1, 1, 1, 0, 0]
[1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]
[0, 1, 1, 1, 0, 1, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1]


In [5]:
def class_wise_acc(y_actual, y_predicted):
    total_p = 0
    total_n = 0
    TP=0
    TN=0
    for i in range(len(y_predicted)):
        if y_actual[i]==1:
            total_p = total_p+1
            if y_actual[i]==y_predicted[i]:
               TP=TP+1
        if y_actual[i]==0:
            total_n=total_n+1
            if y_actual[i]==y_predicted[i]:
               TN=TN+1
    return(TP/total_p, TN/total_n)

class_1_acc_asian, class_0_acc_asian = class_wise_acc(y_actual_asian, y_predicted_asian)
class_1_acc_african_american, class_0_acc_african_american = class_wise_acc(y_actual_african_american, y_predicted_african_american)

print (f"Category 1 Toxic Comments accuracy for Asians = {class_1_acc_asian}")
print (f"Category 2 Non Toxic Comments accuracy for Asians = {class_0_acc_asian}")
print (f"Category 3 Toxic Comments accuracy for African-Americans = {class_1_acc_african_american}")
print (f"Category 4 Non Toxic Comments accuracy for African-Americans = {class_0_acc_african_american}")

Category 1 Toxic Comments accuracy for Asians = 0.6153846153846154
Category 2 Non Toxic Comments accuracy for Asians = 0.5333333333333333
Category 3 Toxic Comments accuracy for African-Americans = 0.6363636363636364
Category 4 Non Toxic Comments accuracy for African-Americans = 0.9090909090909091


# **Final Insights**

If we compare how close the accuracy values are for both Asians and African-Americans for the Toxic Comments category, we can see that they are roughly close. This means that if a comment is toxic, the model can predict it's toxic with a 61-63% acccuracy.

But what is troubling here with the Perspective API usage is the difference in accuracies for the Non-Toxic catgeories for Asians and African-Americans. For African-Americans, the accuracy is at a steep 90%, but for Asians, it's only 53%. This difference indicates that the model is able to recognize toxicity for hate speech against African Americans over that against Asians.



# **Detailed Evaluation for Asian American Group**

**To identify why there is a discrepancy just for the Non-Toxic comments for Asian Americans**

It was surprising to see 61.5% accuracy for the Toxic category for Asian Americans, since most if not all tweets in the dataset were offensive. But, the Perspective API model was not able to detect this due to any subtleties in language, or. Something that could've helped us to delve deeper into understanding the real reasons as to why the accuracies were lower is if the Perspective API model had algorthmic transparency. If us users were given access to the data that the API model was trained/tested on would give us more context.  


The accuracy for the Non-Toxic comments for Asians indicated to only be about 50% correct, because of how some tweeters decided to use swear words and profanity to express their frustration about the racist situation post COVID-19 outbreak in the US but they weren't trying to demonstrate any hate towards Asians. This shows that the Perspective API model is weak in understanding tone and mood of the semantics used in statements. In these tweets, they were using toxic vocabulary, like profane language, due to which the API automatically flagged them as "Toxic" but are really not and rather are showing support and defending the Asian American population on Twitter.



# **Conclusion + Limitations + Scope for Further Research**

Based on our study conducted, the analysis we performed, and insights we drew from it, we can conclude that the scope for th ePerspective API being able to identify or demarcate differences in semantics, mood, message, and overall tone of the claim that is inputted is quite limited. This was shown by how the accuracy values were extremely different for the Non-Toxic category in specific for Asian Americans and African Americans. This bias in identifying toxicity for African Americans more than Asian Americans propels us to think about the potential limitations of the model.

As identified above, the main limitation of the Perspective API model is that we don't have much context on the dataset it was trained on. There is not much Algorithmic Transparency here, as us users do not have access to the dataset that the model was trained on. This context would help us understand why there was a difference of 40% of accuracy between the Non-toxic accuracy for both marginalized groups.

The scope for further research is large here, since this study uncovered great insights into racial differenciation between two different set of marginalized groups in social media. Apart from increasing the sample size, one of the opportunities that can be explored is considering other social media other than Twitter like Facebook, Instagram, and etc. Potentially, we can compare the results obtained for Tweets vs. Facebook posts and understand whether there are more factors affecting it. Another method to extend this research is considering the political side of it, since political arguments are very common in the disucssion of discrimination against marginalized groups in the US. These methods would both equally be interesting and intriguing to explore.